### PR TITLE
Mark / install config.json

### PR DESCRIPTION
**Pull request by markccchiang**
I modify the file "carta/carta.pro", add two files "carta/config/config.json" and "carta/config/config.pro" in this branch. So the main configuration file will automatically install in the Carta build directory as the default set.

### CODE REVIEW

File: carta/carta.pro [hunk 1]
```diff
@@ -14,12 +14,12 @@ dbg( "===---------------------------------------------====")
 
 TEMPLATE = subdirs
 SUBDIRS += cpp
-SUBDIRS += scriptedClient
-SUBDIRS += scripts
+SUBDIRS += config
+# SUBDIRS += scriptedClient
```
Comment (grimmerk) [inline carta/carta.pro:18]: Commenting these two lines means they did not do anything before?

**Comment by markccchiang** [comment]
Yes, I think so.

2017-04-10 11:49 GMT+08:00 Grimmer <notifications@github.com>:

**Comment by grimmerk** [comment]
@markccchiang , @ajm-asiaa 
When releasing, "qtDecorations" should be  "false".

How to tell it is building in release mode:
1. CARTA_BUILD_TYPE = release

**Comment by ajm-ska** [comment]
I'm not sure if this is working, at least when compiling the old way. 
I seem to end up with a file called "empty.lprof" of 0 bytes in the "~/cartabuild/build/cpp/desktop/desktop.app/Contents/Resources/ folder"

Do you have that Mark?

Of course I know I could manually copy the config.json there from ~/cartabuild/build/config/ as an extra step.

**Comment by ajm-ska** [comment]
Oh I see. An old version of the code also has a "empty.lprof" file in the desktop.app/Contents/Resources/ folder.

So I guess we do need a post build step to copy the config.json into there?

I see there is a difference in the contents of the build folder for this commit and an older version of the source code:
This commit: /build contains Makefile config/ cpp/
Older version: /build contains Makefile cpp/ scriptedClient/ scripts/

**Comment by ajm-ska** [comment]
One more question Mark.
Which carta.sh script do you use?

**Comment by markccchiang** [comment]
Hi Anthony,

We don't need to copy config.json to the directory
CARTAVis/build/config/config.json. "qmake" will do this step automatically,
since I have modified the file CARTAVis/carta/carta.pro in this branch.
BTW, the carta.sh script I used is the same with develop branch.

2017-04-13 13:24 GMT+08:00 Anthony Moraghan <notifications@github.com>:

File: carta/carta.pro [hunk 2]
```diff
@@ -14,12 +14,12 @@ dbg( "===---------------------------------------------====")
...
+# SUBDIRS += scripts
 
 OTHER_FILES += readme.txt uncrustify.cfg
 
-
 # make sure user did not specify debug version...
 # it's probably harmless, but that's probably not what the user wanted
 CONFIG(debug,debug|release) {
```
Comment (ajm-ska) [inline carta/carta.pro]: test

File: carta/config/config.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+        "$(APPDIR)/../plugins",
+        "$(APPDIR)/../../../../plugins"
+    ],
+    "disabledPlugins" : ["casaCore-2.0.1"],
+    "plugins": {
+        "PCacheSqlite3" : {
+            "dbPath": "/tmp/pcache.sqlite"
+        }
+    },
+    "qtDecorations" : "true"
```
Comment (ajm-ska) [inline carta/config/config.json:13]: tedst2

**Review by ajm-ska — COMMENTED** [review]
test3

**Comment by ajm-ska** [comment]
Thanks Mark. It seems to me that when we run "carta.sh", if ~/.cartavis/config.json is not present, then it will run setupcartavis.sh which will copy it there.

/CARTAvis/carta/scripts/legacy/carta.sh:
```
if [ ! -d $HOME/CARTA/Images/CubesTest -o\
     ! -d $HOME/.cartavis/log  -o\
     ! -f $HOME/.cartavis/config.json -o\
     ! -d $HOME/CARTA/snapshots/data ]; then
    $dirname/setupcartavis.sh 2>&1  > /dev/null
fi
```

/CARTAvis/carta/scripts/setupcartavis.sh
```
# check that config.json exists
if [ ! -f $HOME/.cartavis/config.json ]; then
    echo "copying config.json file to  $HOME/.cartavis directory..."
    cp $dirname/config/config.json $HOME/.cartavis
fi
```

Do we need to stop this behaviour?

**Comment by markccchiang** [comment]
If Grimmer has no other opinions, I think we can disable this behavior.

2017-04-13 14:00 GMT+08:00 Anthony Moraghan <notifications@github.com>:

**Comment by ajm-ska** [comment]
It doesn't seem to work for me without config.json in the usual place.
```
Parsing global settings from "/Users/ajm/.cartavis/config.json"
Could not open main config file from "/Users/ajm/.cartavis/config.json"
```

So for you it reads config.json from /Contents/Resources/config/config.json ?

**Comment by markccchiang** [comment]
It seems the current develop branch does not merge my previous pull
request:

https://github.com/CARTAvis/carta/pull/46/files

which can read the config.json from the default directory
carta/config/config.json. I will check with Grimmer.


2017-04-13 14:30 GMT+08:00 Anthony Moraghan <notifications@github.com>:

**Comment by ajm-ska** [comment]
Hi Mark. OK. I combined pull request 46 and 47 together. 
It is going back through too many directories after we package CARTA:

> Parsing global settings from "/private/tmp/Carta.app/Contents/MacOS/../../../../../config/config.json"
> Could not open main config file from "/private/tmp/Carta.app/Contents/MacOS/../../../../../config/config.json"

For Mac it would only need to go back one directory and then into 'Resources'. For example, it could be /Carta.app/Contents/Resources/config/config.json

So I propose to change Line62 of CmdLine.cpp to
```
 info.m_configFilePath = QCoreApplication::applicationDirPath() + "/../Resources/config/config.json";
```
I tried it and it works.

Similarly for the packaged Linux case, Line 60 of CmdLine.cpp should be
```
info.m_configFilePath = QCoreApplication::applicationDirPath() + "/../etc/config/config.json";
```

File: carta/cpp/core/CmdLine.cpp [hunk 4]
```diff
@@ -45,11 +48,20 @@ ParsedInfo parse(const QStringList & argv)
...
         info.m_configFilePath = QDir::homePath() + "/.cartavis/config.json";
     }
 
+    // if the config file was not specified neither through command line or environment
+    // assign a default value (search the config file under the carta build directory second)
+    if (!QFile::exists(info.m_configFilePath)) {
+#ifdef Q_OS_LINUX
+        info.m_configFilePath = QCoreApplication::applicationDirPath() + "/../../config/config.json";
+#else
+        info.m_configFilePath = QCoreApplication::applicationDirPath() + "/../../../../../config/config.json";
```
Comment (ajm-ska) [inline carta/cpp/core/CmdLine.cpp]: info.m_configFilePath = QCoreApplication::applicationDirPath() + "/../Resources/config/config.json";

File: carta/cpp/core/CmdLine.cpp [hunk 5]
```diff
@@ -45,11 +48,20 @@ ParsedInfo parse(const QStringList & argv)
...
+    // assign a default value (search the config file under the home directory first)
     if( info.m_configFilePath.isEmpty()) {
         info.m_configFilePath = QDir::homePath() + "/.cartavis/config.json";
     }
 
+    // if the config file was not specified neither through command line or environment
+    // assign a default value (search the config file under the carta build directory second)
+    if (!QFile::exists(info.m_configFilePath)) {
+#ifdef Q_OS_LINUX
+        info.m_configFilePath = QCoreApplication::applicationDirPath() + "/../../config/config.json";
```
Comment (ajm-ska) [inline carta/cpp/core/CmdLine.cpp:60]: info.m_configFilePath = QCoreApplication::applicationDirPath() + "/../etc/config/config.json";

**Review by ajm-ska — CHANGES_REQUESTED** [review]
These changes should be made so that it will work once CARTA is packaged.

File: carta/cpp/core/CmdLine.cpp (hunk 4 above)
Comment (markccchiang) [reply carta/cpp/core/CmdLine.cpp]: Thanks Anthony! I have made changes in `CmdLine.cpp`. Please check if it works in CARTA packaged program.

File: carta/cpp/core/CmdLine.cpp (hunk 5 above)
Comment (markccchiang) [reply carta/cpp/core/CmdLine.cpp:60]: Thanks Anthony! I have made changes in `CmdLine.cpp`. Please check if it works in CARTA packaged program.

**Review by ajm-ska — APPROVED** [review]

**Comment by ajm-ska** [comment]
I checked on Mac and CentOS7 versions. This works fine.
 
I can just copy the config.json to the correct location in a packaging step.

Maybe next time when we are improving the carta.sh script, we can remove or comment out lines 38 to 42 of the setupcartavis.sh script in order to stop it copying config.json to ~/.cartavis/

> check that config.json exists
> if [ ! -f $HOME/.cartavis/config.json ]; then
>	echo "copying config.json file to  $HOME/.cartavis directory..."
>	cp $dirname/config/config.json $HOME/.cartavis
> fi 

Meanwhile, I can just comment out those lines when packaging.

**Pull request merged**